prow/plugins/lint/lint.go

```diff
@@ -0,0 +1,199 @@
+}
+
+func handle(ghc githubClient, gc *git.Client, log *logrus.Entry, ic github.IssueCommentEvent) error {
+	// Only handle open PRs and new requests.
+	if ic.Issue.State != "open" || !ic.Issue.IsPullRequest() || ic.Action != "created" {
+		return nil
+	}
+	if !lintRe.MatchString(ic.Comment.Body) {
+		return nil
+	}
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+
+	// Find all modified go packages.
+	changes, err := ghc.GetPullRequestChanges(org, repo, ic.Issue.Number)
+	if err != nil {
+		return err
+	}
+
+	modifiedFiles := make(map[string]string)
+	for _, change := range changes {
+		if filepath.Ext(change.Filename) == ".go" {
+			modifiedFiles[change.Filename] = change.Patch
+		}
+	}
+	if len(modifiedFiles) == 0 {
+		return nil
+	}
+
+	r, err := gc.Clone(ic.Repo.FullName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Clean(); err != nil {
+			log.WithError(err).Error("Error cleaning up repo.")
+		}
+	}()
+
+	if err := r.CheckoutPullRequest(ic.Issue.Number); err != nil {
+		return err
+	}
+
+	l := new(lint.Linter)
+	var problems []lint.Problem
+	for f, patch := range modifiedFiles {
+		src, err := ioutil.ReadFile(filepath.Join(r.Dir, f))
+		if err != nil {
+			return err
+		}
+		ps, err := l.Lint(f, src)
+		if err != nil {
+			return err
+		}
+		al, err := addedLines(patch)
+		if err != nil {
+			return err
+		}
+		for _, p := range ps {
+			if al[p.Position.Line] {
+				problems = append(problems, p)
+			}
+		}
+	}
+	var response string
+	if len(problems) == 0 {
+		response = "no lint warnings"
+	} else {
+		response = fmt.Sprintf("%d warnings:\n\n", len(problems))
+		var warnings []string
+		for _, p := range problems {
+			warnings = append(warnings, fmt.Sprintf("%s:%d: %s", p.Position.Filename, p.Position.Line, p.Text))
+		}
+		response += strings.Join(warnings, "\n")
+	}
+
+	return ghc.CreateComment(org, repo, ic.Issue.Number, plugins.FormatICResponse(ic.Comment, response))
+}
+
+// addedLines returns a set of line numbers that were added in the patch.
+// https://www.gnu.org/software/diffutils/manual/diffutils.html#Detailed-Unified
+// GitHub omits the ---/+++ lines since that information is in the
+// PullRequestChange object.
+func addedLines(patch string) (map[int]bool, error) {
+	result := make(map[int]bool)
+	lines := strings.Split(patch, "\n")
+	for i := 0; i < len(lines); i++ {
+		_, oldLen, newLine, newLen, err := parseHunkLine(lines[i])
+		if err != nil {
+			return nil, err
+		}
+		oldAdd := 0
+		newAdd := 0
+		for oldAdd < oldLen || newAdd < newLen {
+			i++
+			if i >= len(lines) {
+				return nil, fmt.Errorf("invalid patch: %s", patch)
+			}
+			switch lines[i][0] {
+			case ' ':
+				oldAdd++
+				newAdd++
+			case '-':
+				oldAdd++
+			case '+':
+				result[newLine+newAdd] = true
+				newAdd++
+			default:
+				return nil, fmt.Errorf("bad line in patch: %s", lines[i])
+			}
+		}
+	}
+	return result, nil
+}
+
+var hunkRe = regexp.MustCompile(`@@ -([\d]+),?([\d]+)? \+([\d]+),?([\d]+)? @@.*`)
```

Comment:
Can you add a comment with some examples that this regexp is supposed to match? Parsing regexps is not fun. :)